h2o-algos/src/main/java/hex/modelselection/ModelSelectionUtils.java

```diff
@@ -298,6 +299,99 @@ public static String[] extractPredictorNames(ModelSelectionModel.ModelSelectionP
+            coeffNames.remove(interceptIndex);
+        }
+        // grab min z-values for numerical and categorical columns
+        PredNameMinZVal numericalPred = findNumMinZVal(numPredNames, zValList, coeffNames);
+        PredNameMinZVal categoricalPred = findCatMinZVal(catPredNames, zValList, coeffNames);
+        
+        // choose the min z-value from numerical and categorical predictors and return its index in predNames
+        if (categoricalPred._minZVal >= 0 && numericalPred._minZVal >= 0) {
+            if (categoricalPred._minZVal < numericalPred._minZVal) {
+                catPredNames.remove(catPredNames.indexOf(categoricalPred._predName));
+                return predNames.indexOf(categoricalPred._predName);
+            } else {
+                numPredNames.remove(numPredNames.indexOf(numericalPred._predName));
+                return predNames.indexOf(numericalPred._predName);
+            }
+        } else if (categoricalPred._minZVal >= 0) { // categorical pred has minimum z-value
+            catPredNames.remove(catPredNames.indexOf(categoricalPred._predName));
+            return predNames.indexOf(categoricalPred._predName);
+        } else {    // numerical pred has minimum z-value
+            numPredNames.remove(numPredNames.indexOf(numericalPred._predName));
+            return predNames.indexOf(numericalPred._predName);
+        }
+    }
+    
+    public static PredNameMinZVal findNumMinZVal(List<String> numPredNames, List<Double> zValList, List<String> coeffNames) {
+        double minNumVal = -1;
+        String numPredMinZ = null;
+        if (numPredNames != null && numPredNames.size() > 0) {
+            List<Double> numZValues = new ArrayList<>();
+            for (String predName : numPredNames) {
+                int eleInd = coeffNames.indexOf(predName);
+                double oneZValue = zValList.get(eleInd);
+                if (Double.isNaN(oneZValue)) {
+                    zValList.set(eleInd, 0.0);
+                    numZValues.add(0.0);    // NaN corresponds to coefficient of 0.0
+                } else {
+                    numZValues.add(oneZValue);
+                }
+            }
+            minNumVal = numZValues.stream().min(Double::compare).get(); // minimum z-value of numerical predictors
+            numPredMinZ = numPredNames.get(numZValues.indexOf(minNumVal));
+        }
+        return new PredNameMinZVal(numPredMinZ, minNumVal);
+    }
+    
+    public static PredNameMinZVal findCatMinZVal(List<String> catPredNames, List<Double> zValList, List<String> coeffNames) {
+        double minCatVal = -1;
+        String catPredMinZ = null;
+        if (catPredNames != null && catPredNames.size() > 0) {
+            minCatVal = Double.MAX_VALUE;
+            for (String catName : catPredNames) {
+                List<Double> catZValues = new ArrayList<>();
+                for (String coeffName : coeffNames) {
+                    if (coeffName.startsWith(catName)) {
```

Comment:
That is a great point!

Reply:
@tomasfryda  I have come up with a different way to do this. 

I basically generated a HashMap of cat predictor names and a List<String> of cat predictor name + "." + all cat levels.  

Then, as I look at each categorical column, I will bring up the list of cat predictor name + "." + all cat levels which are the coefficient names.  There will be one less coefficients than all the levels.  Then, I will collect all the z-values of the coefficients and then grab the maximum abs(z-values).  Hope this one works.  Please let me know.

Reply:
Thank you @wendycwong. This should behave more correctly but there could still be an issue - consider variables a=["b.c", "d"] and a.b=["c", "d"] - in this case we encode the variables and end up with two predictors with the name `a.b.c`. We could detect it and fail in this case.

I'm not sure but I think the `missing(NA)` "category" isn't present in the `domains` would you make sure of it? Basically, If you have a missing value in categorical column it gets represented as `category_column.missing(NA)` on the client so I'd assume this happens on the backend as well (and it is not present in the domains AFAIK).

And the last thing I have some doubts is the usage of `max(abs(z-value))` when we get rid of the column we get rid of all the encoded categories so I'd assume there would be `sum` instead of `max` but maybe I'm missing something. Just intuitively if I'd have a column with hundreds of categories and each of those categories have relatively low z-value I could throw away a column that would in total bring most of the information (the information would be just distributed to multiple levels).

Reply:
@tomasfryda Regarding the first point, it is very interesting because we will be having duplicated coefficient names and it can be very confusing.  I think failing it is a good way to hand it.  I will add it.

Regarding NAs, I have not given it one iota of though.  Let me add it.

For the last question, I just need the magnitude of abs(z-value).  Basically, I am trying to do the following:

at first iteration, I will build a glm model and decide which predictor to eliminate for the next iteration.  The criteria to remove a predictor is to look at its p-values.  The larger the p-value, the more insignificant the predictor is.  Large p-value implies smaller abs(z-value).  Hence, I am looking at the predictor with the smallest abs(z-value) to remove.

at second iteration, I will repeat what is going on in iteration one except with one less predictor.  Again, I am trying to remove another predictor.

The iterations continue until there is no more to remove or min_predictor_num is reached.

Things get complicated when there is categorical columns.  Instead of one coefficient, it has many.  Therefore there are many z-values to consider as well.  I don't want to remove a categorical column as long as if it has one significant level, meaning it has a small p-value (or large abs(z-value)).  Therefore, I will collect all its z-values, and take the max(abs(z-value)).  If this value if smaller than other predictors, I will then remove the categorical column.

Reply:
@wendycwong Yes, things are complicated with the categorical column.

I assume the p-values correspond to the hypothesis that the coefficient is non-zero for the numerical columns.

If that's the case I would think that for categorical columns we should do the same but that's the problem if we use OHE/treatment contrast/dummy encoding then we have one "reference category" that is "active" when all encoded columns are not so we cannot really test for it - that's the first issue. I did some experiments (using R's lm) and this doesn't seem to be an issue - the "importance" (t-value) moves to the rest of the categories. 

Another one is, I think, that we should be looking at it as a multivariate problem and using statistics like Hotelling's T^2 to find out if the coefficients of all the encoded columns are non-zero. (But since we use z-values we might as well use multinomial normal distribution.)

The way I think about it is that we have hypothesis that a categorical column is useless so when we encode it the hypothesis is transformed to a multivariate normal distribution with mean (0,0,...,0) and if we assume same dispersion across the dimensions we have a nice symmetric distribution. When we look at all the neighborhoods of the same probability density (z-values) we have a nice hyper-sphere.  To make it simple lets assume just 2 dimensions, in this case we have a circle with a center in (0, 0). To simplify further let's assume it is a unit circle (`x^2+y^2 = 1`). And now imagine two cases:
1. one category is important and the other one is not - let's put it at the boundary of the unit circle (1, 0)
2. both categories are important => (1, 1)

Now when we would take the max of the z-values we would end in both cases with 1 but in the second case we know that that case is less likely as it lies outside of the unit circle. I made a simple image - axis correspond to 2 categories, the `x` marks depict the z-value of the 2 cases I mentioned earlier.  The blue x (case 1) and green x (case 2) have the same `max(z-value)` but as you can see on the red gradient (the probability density of the cat. column being useless) in the multivariate case the green x is far less likely to be (0,0) than the blue x.
![ball](https://user-images.githubusercontent.com/61695433/165936989-229ee63d-dac8-4456-8dd7-17f61b7ad61c.png)

Note that I'm not a statistician so I might have written a complete nonsense.

Reply:
@tomasfryda I completely changed the approach I use to map the coefficient names to column names and this should resolve your concern.  

It turns out that the dataframe used to train a model is the dinfo._adaptedFrame.  I can get the column names in order by calling glmModel.names.  Next, the coefficient names are generated from the dinfo._adaptedFrame.  The categorical columns are first and the number of levels of the cat columns are recorded in dinfo._catOffsets.  Hence, I can tell which coefficient corresponds to which coefficient names and which coefficient names correspond to which column (predictor) by using the glmModel.names and the dinfo._catOffsets.

For example, if I have a dinfo._adaptedFrame with two cat columns and first one has three levels and second one has two levels, I will have:
names: col1, col2;
catOffsets: 0, 3, 5 (offset by one).

The coefficient names generated will be:
col1_level0
co1_level1
col1_level2
col2_level0
col2_level1.

The zvalues generated will correspond to the coefficient names.  There is no mapping between names and levels and everything should work.

Reply:
@tomasfryda

For numerical columns, p-value will measure how important this predictor is.  A small p-value shows that this predictor is important and contribute to the model while a large p-value shows that this predictor is not useful.

A coefficient of zero basically says this predictor does not contribute to the model.  You can remove it and not use it.

My reasoning to keep the categorical column if only one of its level has small p-value is because that particular categorical level is import to the model.  It is a very simple heuristic.